concordia/views.py

```diff
@@ -1404,7 +1404,9 @@ def submit_tags(request, *, asset_pk):
 
     for tag in existing_user_tags:
         if tag not in all_submitted_tags:
-            user_tags.tags.remove(tag)
+            for collection in asset.userassettagcollection_set.all():
+                if tag in collection.tags.all():
```

Comment:
Couldn't this be simplified to `collection.tags.remove(tag)` to avoid the extra queries needed to load the entire tag set? If you computed the tags as a set difference (i.e. something like `set(all_submitted_tags).difference(existing_user_tags)`) you could then make a single call `collection.tags.remove(*all_removed_tags)`.

Similarly, rather than iterating over every collection individually you should be able to do this as a single query by hitting `UserAssetTagCollection` directly — something like `UserAssetTagCollection.tags.through.objects.filter(tag__in=removed_tags, userassettagcollection__asset=asset).delete()` but I'd need to test that code.

Reply:
Great suggestions, thank you. I don't think the "tag__in" filter will work (since the field name is "tags", not "tag"), but I should be able to update the branch with the other code changes.